src/main/java/com/limechain/trie/structure/nibble/Nibbles.java

```diff
@@ -1,86 +1,78 @@
 import java.util.stream.Stream;
-
-// TODO:
-//  Should this implement List or just Iterable... or maybe extend List<Nibble>?
-//  This current version contains a huge amount of boilerplate,
-//  but it's what I've reached as a most convenient option to use outside.
-//  Basically a List<Nibble> with additional methods.
-
-// TODO:
-//  As a List<Nibble> implementor, Nibbles is currently mutable, which I severely dislike,
-//  since this mandates the developer to think about where to copy and where passing by reference is sufficient.
+import java.util.stream.StreamSupport;
 
 /**
- * Convenience wrapper for any 'sequence of Nibble'-like structure
+ * Convenience wrapper for any 'sequence of Nibble'-like structure.
+ * It's immutable and its public static factory methods eagerly copy passed data (nibbles) to obtain ownership.
  */
-public class Nibbles implements List<Nibble>, RandomAccess, Comparable<List<Nibble>> {
-    public static final Nibbles EMPTY = new Nibbles(new byte[]{});
-
+public class Nibbles implements Iterable<Nibble>, RandomAccess, Comparable<Iterable<Nibble>> {
     /**
-     * Don't mutate! :D
+     * A sequence of zero nibbles, i.e. empty
      */
-    public static final Nibbles ALL = new Nibbles(Nibble.all());
+    public static final Nibbles EMPTY = new Nibbles(List.of());
 
-    public static Stream<Nibble> all() {
-        return ALL.stream();
-    }
+    /**
+     * A sequence of all nibbles in order, from 0 to F
+     */
+    public static final Nibbles ALL = Nibble.all().collect(NibblesCollector.toNibbles());
 
     private final List<Nibble> nibbles;
 
-    // TODO: Maybe convert all those overloaded constructors to static generational methods
-    public Nibbles(Nibble nibble) {
-        this(List.of(nibble));
+    public static Nibbles of(Nibble nibble) {
+        return Nibbles.of(List.of(nibble));
+    }
+
+    public static Nibbles of(Nibble[] nibbles) {
+        return Nibbles.of(Arrays.asList(nibbles));
     }
 
-    public Nibbles(Nibble[] nibbles) {
-        this(Arrays.asList(nibbles));
+    public static Nibbles of(Collection<Nibble> nibbles) {
+        return new Nibbles(new ArrayList<>(nibbles));
     }
 
-    public Nibbles(byte[] nibbles) {
-        this(Arrays.stream(ArrayUtils.toObject(nibbles)).map(Nibble::fromByte).toList());
+    public static Nibbles of(Stream<Nibble> nibbles) {
+        return Nibbles.of(nibbles.iterator());
     }
 
-    public Nibbles(Iterable<Nibble> nibbles) {
-        this(nibbles.iterator());
+    public static Nibbles of(Iterable<Nibble> nibbles) {
+        return Nibbles.of(nibbles.iterator());
     }
 
-    public Nibbles(Stream<Nibble> nibbles) {
-        this(nibbles.iterator());
+    public static Nibbles of(Iterator<Nibble> nibbles) {
+        return new Nibbles(IteratorUtils.toList(nibbles));
     }
 
-    public static Nibbles of(int... nibbles) {
-        return new Nibbles(Arrays.stream(nibbles).mapToObj(Nibble::fromInt));
+    Nibbles(List<Nibble> nibbles) {
```

Comment:
You can make the constructor private and create a public `of(List)` method to keep the class consistent.